### DISCUSSION TITLE
How to set the pollutant concentration for a node using the pyswmm library?

### DISCUSSION THREAD

**Discussion by jack06221** [pyswmm/pyswmm#505] [Q&A] [answered]
I have tried methods like node.set_pollutant(), sim.node_set_pollutant(), node.pollutant_concentration['PollutantName'] = 10, node.node_set_pollutant('TSS', 0), but none of them worked.

Here is my code：
'''
PySWMM and StormReactor Code Example A
Author: Brooke Mason
Version: 1
Date: December 21, 2022
'''
# Import libraries
from pyswmm import Simulation, Nodes
from StormReactor import waterQuality
import matplotlib.pyplot as plt

# Create lists to save TSS results
J1_TSS = []
J2_TSS = []
J3_TSS = []
J4_TSS = []
J5_TSS = []
J6_TSS = []
J7_TSS = []
J8_TSS = []
J9_TSS = []
J10_TSS = []
J11_TSS = []
J12_TSS = []
J13_TSS = []
OUT_TSS = []

# Initalize SWMM simulation
with Simulation(r'Example1_WQ.inp') as sim:
    # Node information
    J1Node = Nodes(sim)['J1']
    J2Node = Nodes(sim)['J2']
    J3Node = Nodes(sim)['J3']
    J4Node = Nodes(sim)['J4']
    J5Node = Nodes(sim)['J5']
    J6Node = Nodes(sim)['J6']
    J7Node = Nodes(sim)['J7']
    J8Node = Nodes(sim)['J8']
    J9Node = Nodes(sim)['J9']
    J10Node = Nodes(sim)['J10']
    J11Node = Nodes(sim)['J11']
    J12Node = Nodes(sim)['J12']
    J13Node = Nodes(sim)['J13']
    OUTNode = Nodes(sim)['OUT']

    # 设置节点浓度
    J1Node.set_pollutant('TSS', 10)
    sim.node_set_pollutant(J1Node, 'TSS', 10)
    J1Node.pollutant_concentration['PollutantName'] = 10
    J2Node.node_set_pollutant('TSS', 0)
    # J3Node.node_set_pollutant('TSS', 0)
    # J4Node.node_set_pollutant('TSS', 0)
    # J5Node.node_set_pollutant('TSS', 0)
    # J6Node.node_set_pollutant('TSS', 0)
    # J7Node.node_set_pollutant('TSS', 0)
    # J8Node.node_set_pollutant('TSS', 0)
    # J9Node.node_set_pollutant('TSS', 0)
    # J10Node.node_set_pollutant('TSS', 0)
    # J11Node.node_set_pollutant('TSS', 0)
    # J12Node.node_set_pollutant('TSS', 0)
    # J13Node.node_set_pollutant('TSS', 0)
    # OUTNode.node_set_pollutant('TSS', 0)

    # Initialize StormReactor
    # WQ = waterQuality(sim, config)
    WQ = waterQuality(sim)

    # Launch a simulation
    for step in sim:
        # Update water quality each simulation step
        WQ.updateWQState()
        # Get and save TSS concentrations
        J1_TSS.append(J1Node.pollut_quality['TSS'])
        J2_TSS.append(J2Node.pollut_quality['TSS'])
        J3_TSS.append(J3Node.pollut_quality['TSS'])
        J4_TSS.append(J4Node.pollut_quality['TSS'])
        J5_TSS.append(J5Node.pollut_quality['TSS'])
        J6_TSS.append(J6Node.pollut_quality['TSS'])
        J7_TSS.append(J7Node.pollut_quality['TSS'])
        J8_TSS.append(J8Node.pollut_quality['TSS'])
        J9_TSS.append(J9Node.pollut_quality['TSS'])
        J10_TSS.append(J10Node.pollut_quality['TSS'])
        J11_TSS.append(J11Node.pollut_quality['TSS'])
        J12_TSS.append(J12Node.pollut_quality['TSS'])
        J13_TSS.append(J13Node.pollut_quality['TSS'])
        OUT_TSS.append(OUTNode.pollut_quality['TSS'])

# Plot TSS concentrations
plt.plot(J1_TSS, ':', label="J1_TSS")
plt.plot(J2_TSS, ':', label="J2_TSS")
plt.plot(J3_TSS, ':', label="J3_TSS")
plt.plot(J4_TSS, ':', label="J4_TSS")
plt.plot(J5_TSS, ':', label="J5_TSS")
plt.plot(J6_TSS, ':', label="J6_TSS")
plt.plot(J7_TSS, ':', label="J7_TSS")
plt.plot(J8_TSS, ':', label="J8_TSS")
plt.plot(J9_TSS, ':', label="J9_TSS")
plt.plot(J10_TSS, ':', label="J10_TSS")
plt.plot(J11_TSS, ':', label="J11_TSS")
plt.plot(J12_TSS, ':', label="J12_TSS")
plt.plot(J13_TSS, ':', label="J13_TSS")
plt.plot(OUT_TSS, ':', label="OUT_TSS")

plt.ylim(0, 100)
plt.xlabel("Time")
plt.ylabel("TSS (mg/L)")
plt.legend()
plt.show()

**Comment by jack06221** [comment]
Here is my SWMM file and code. 
[1.zip](https://github.com/pyswmm/swmmio/files/13858857/1.zip)

**Reply by bemcdonnell** [reply]
@jack06221 i transferred this to pyswmm since I believe you’re trying to interact with the sim. @bemason do you mind giving a bit of advice here? I have not spent much time using pollutants workflow :-)

**Reply by bm-hyfi** [reply]
I am currently on PTO and responding via my cell. @abhiramm7 do you have some time to give them some advice? otherwise I can probably look at this over the weekend. sorry for the delay!

**Comment by bemcdonnell** [comment, marked as answer]
@jack06221 If youre trying to set a concentration I think you’d need to call:

`
J2Node.pollut_quality = ('TSS', 100)
`

based on the documentation- it looks like the setter accepts a tuple. Which means, if you have multiple pollutants you’d need to set each one. 

Also, I believe you’ll need to set this during your simulation (unless you’re trying to set the initial value as you have it now).

**Reply by jack06221** [reply]
If I want to set a time series for a pollutant to change its concentration at different times, how should I proceed?

**Reply by bemcdonnell** [reply]
@jack06221 Are you comfortable writing a small python function to interpolate your timeseries values between samples? you could use PySWMM to get the `sim.current_time` datetime values and sample from your timeseries.  Just and idea

**Comment by karosc** [comment]
One of my issues was that a node needs to be specified in the swmm treatment section for pollutant setter to work (https://github.com/OpenWaterAnalytics/swmm-python/pull/94#issuecomment-1013951491)

**Reply by bemcdonnell** [reply]
@karosc should we consider modifying the API to insert these changes on-demand? We could raise this issue in SWMM
